realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java

```diff
@@ -448,15 +459,26 @@ private boolean categorizeField(Element element) {
             if (!categorizeIndexField(element, field)) { return false; }
         }
 
-        if (isRequiredField(field)) {
-            categorizeRequiredField(element, field);
+        // @Required annotation of RealmList field only affects its value type, not field itself.
+        if (Utils.isRealmList(field)) {
+            // We only check @Required annotation. @org.jetbrains.annotations.NotNull annotation should not affect nullability of the list values.
+            if (!hasRequiredAnnotation(field)) {
```

Comment:
It is a bit tricky. If going solely by the Java definition, there is no way to prevent `null` values in that list. In Kotlin you could do `val list : RealmList<String?>` . Ideally, I would really like to avoid having to add a new annotation just for this.

Reply:
But yes, we need to consider what happens if (when) we allow the list itself to be null

Reply:
After sleeping a little on this, I think I fall on the side that adding a new `@ElementsRequired` annotation is overkill. To me the value of allowing a `list` to be null is so low that it might never happen, so trying to optimize for that now would be a bit much.

It also means less work for us at this point and we can cross that bridge when we get there.

For Kotlin this is rather annoying though, because `RealmList<String?>` is not visible to our annotation processor, but instead encoded in a special @Metadata annotation. 

This library here: https://github.com/Takhion/kotlin-metadata can help parse that, but that is definitely out of scope for 4.0, so I would just create an issue for that and we should document this caveat on our website.

What do you think @realm/java ?

Reply:
OK, if most people think we are not going to have nullable list field :P I am fine then.